### PR TITLE
validation temp

### PR DESCRIPTION
**Pull request by nikhita**

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+func addConversionFuncs(scheme *runtime.Scheme) error {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go:27]: must be registered in register.go

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/deepcopy.go [hunk 2]
```diff
@@ -24,13 +24,13 @@ func (in *JSONSchemaProps) DeepCopy() *JSONSchemaProps {
 	out := new(JSONSchemaProps)
 	*out = *in
 
-	if in.Default != nil {
+	/*if in.Default != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/deepcopy.go]: these will be updated, right?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go (hunk 1 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go:27]: oh, I see. should solve the problem then. :+1:

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/deepcopy.go (hunk 2 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/deepcopy.go]: yes

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 3]
```diff
@@ -163,7 +161,7 @@ type JSONSchemaProps struct {
 	Type                 string                     `json:"type,omitempty"`
 	Format               string                     `json:"format,omitempty"`
 	Title                string                     `json:"title,omitempty"`
-	Default              interface{}                `json:"default,omitempty"`
+	Default              JSON                       `json:"default,omitempty"`
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: in fact I like this extra type. Much more explicit, also in internal

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go [hunk 4]
```diff
@@ -280,6 +330,16 @@ func Convert_apiextensions_ExternalDocumentation_To_v1beta1_ExternalDocumentatio
 	return autoConvert_apiextensions_ExternalDocumentation_To_v1beta1_ExternalDocumentation(in, out, s)
 }
 
+func autoConvert_v1beta1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s conversion.Scope) error {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go]: this is strange indeed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go [hunk 5]
```diff
@@ -303,34 +360,146 @@ func autoConvert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(in *JS
...
+			if err := Convert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(&val, newVal, s); err != nil {
+				return err
+			}
+			(*out)[key] = *newVal
+		}
+	} else {
+		out.Definitions = nil
 	}
+	out.ExternalDocs = (*apiextensions.ExternalDocumentation)(unsafe.Pointer(in.ExternalDocs))
+	// WARNING: in.Example requires manual conversion: inconvertible types (k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1.JSON vs k8s.io/apiextensions-apiserver/pkg/apis/apiextensions.JSON)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go]: It should find these. We have to investigate why it doesn't.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go [hunk 6]
```diff
@@ -339,8 +508,7 @@ func autoConvert_apiextensions_JSONSchemaProps_To_v1beta1_JSONSchemaProps(in *ap
 	out.Type = in.Type
 	out.Format = in.Format
 	out.Title = in.Title
-	// TODO: Inefficient conversion - can we improve it?
-	if err := s.Convert(&in.Default, &out.Default, 0); err != nil {
+	if err := Convert_apiextensions_JSON_To_v1beta1_JSON(&in.Default, &out.Default, s); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go]: This direction works!

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 7]
```diff
@@ -67,11 +67,17 @@ func convertJSONSchemaProps(in *apiextensions.JSONSchemaProps, out *spec.Schema)
 	out.MinItems = in.MinItems
 	out.UniqueItems = in.UniqueItems
 	out.MultipleOf = in.MultipleOf
-	out.Enum = in.Enum
 	out.MaxProperties = in.MaxProperties
 	out.MinProperties = in.MinProperties
 	out.Required = in.Required
-	out.Example = in.Example
+	out.Default = interface{}(in.Default)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: Is this actually necessary to type-cast?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: Will try removing it.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: and is it what we want? Does it store the type of the json data structure (string, bool, int, float, map, slice) in `Default` or `JSON(...)`?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: needs a comment at least

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go [hunk 8]
```diff
@@ -303,34 +360,146 @@ func autoConvert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(in *JS
...
+	} else {
+		out.Definitions = nil
 	}
+	out.ExternalDocs = (*apiextensions.ExternalDocumentation)(unsafe.Pointer(in.ExternalDocs))
+	// WARNING: in.Example requires manual conversion: inconvertible types (k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1.JSON vs k8s.io/apiextensions-apiserver/pkg/apis/apiextensions.JSON)
 	return nil
 }
 
-// Convert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps is an autogenerated conversion function.
-func Convert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(in *JSONSchemaProps, out *apiextensions.JSONSchemaProps, s conversion.Scope) error {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/zz_generated.conversion.go]: This is removed. :scream:

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 9]
```diff
@@ -27,15 +27,21 @@ import (
...
-		Convert_v1beta_JSON_To_apiextensions_JSON,
+		Convert_v1beta1_JSON_To_apiextensions_JSON,
 	)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
+// Convert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps is an autogenerated conversion function.
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: not sure if this is needed, will try removing and check.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 10]
```diff
@@ -27,15 +27,21 @@ import (
...
+func Convert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(in *JSONSchemaProps, out *apiextensions.JSONSchemaProps, s conversion.Scope) error {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: this should be automatic, shouldn't it? Would not put that here.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: yeah, will remove this.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: I think the error was only in the typo, that should fix the conversion as a whole.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 11]
```diff
@@ -27,15 +27,21 @@ import (
...
+	return autoConvert_v1beta1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(in, out, s)
+}
+
 func Convert_apiextensions_JSON_To_v1beta1_JSON(in *apiextensions.JSON, out *JSON, s conversion.Scope) error {
 	raw, err := json.Marshal(*in)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: interface{}(*in) ?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: Go doesn't complain though.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 12]
```diff
@@ -45,9 +51,11 @@ func Convert_apiextensions_JSON_To_v1beta1_JSON(in *apiextensions.JSON, out *JSO
 	return nil
 }
 
-func Convert_v1beta_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s conversion.Scope) error {
-	if err := json.Unmarshal(in.Raw, *out); err != nil {
+func Convert_v1beta1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s conversion.Scope) error {
+	var raw []byte
+	if err := json.Unmarshal(raw, *out); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]:
```
var i interface{}
Unmarshal(in.raw, &i)
*out = i
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 13]
```diff
@@ -193,7 +190,7 @@ type JSONSchemaProps struct {
 }
 
 type JSON struct {
-	Raw []byte `protobuf:"bytes,1,opt,name=raw"`
+	Raw []byte `json:"raw" protobuf:"bytes,1,opt,name=raw"`
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: Is that json tag needed? We implement JSON marshalling manually.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: will fix.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go [hunk 14]
```diff
@@ -117,3 +117,18 @@ func (s *JSONSchemaPropsOrArray) UnmarshalJSON(data []byte) error {
 	*s = nw
 	return nil
 }
+
+func (s JSON) MarshalJSON() ([]byte, error) {
+	if len(s.Raw) > 0 {
+		return json.Marshal(s.Raw)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: this is one marshal to much. Just do:
```
func (s *JSON) MarshalJSON() ([]byte, error) {
   return s.Raw, nil
}
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go [hunk 15]
```diff
@@ -117,3 +117,18 @@ func (s *JSONSchemaPropsOrArray) UnmarshalJSON(data []byte) error {
...
+	}
+	return []byte("null"), nil
+
+}
+
+func (s *JSON) UnmarshalJSON(data []byte) error {
+	if err := json.Unmarshal(data, &(s.Raw)); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: just: `s.Raw = data`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 16]
```diff
@@ -146,51 +146,51 @@ type CustomResourceDefinitionList struct {
...
-	Not                  *JSONSchemaProps           `json:"not,omitempty"`
-	Properties           map[string]JSONSchemaProps `json:"properties,omitempty"`
-	AdditionalProperties *JSONSchemaPropsOrBool     `json:"additionalProperties,omitempty"`
-	PatternProperties    map[string]JSONSchemaProps `json:"patternProperties,omitempty"`
-	Dependencies         JSONSchemaDependencies     `json:"dependencies,omitempty"`
-	AdditionalItems      *JSONSchemaPropsOrBool     `json:"additionalItems,omitempty"`
-	Definitions          JSONSchemaDefinitions      `json:"definitions,omitempty"`
-	ExternalDocs         *ExternalDocumentation     `json:"externalDocs,omitempty"`
-	Example              JSON                       `json:"example,omitempty"`
+	ID                   string                     `json:"id,omitempty" protobuf:"bytes,1,rep,name=id"`
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: only `rep` for slices like `Enum` or `Required`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 17]
```diff
@@ -146,51 +146,51 @@ type CustomResourceDefinitionList struct {
...
+	Dependencies         JSONSchemaDependencies     `json:"dependencies,omitempty" protobuf:"bytes,32,rep,name=dependencies"`
+	AdditionalItems      *JSONSchemaPropsOrBool     `json:"additionalItems,omitempty" protobuf:"bytes,33,rep,name=additionalItems"`
+	Definitions          JSONSchemaDefinitions      `json:"definitions,omitempty" protobuf:"bytes,34,rep,name=definitions"`
+	ExternalDocs         *ExternalDocumentation     `json:"externalDocs,omitempty" protobuf:"bytes,35,rep,name=externalDocs"`
+	Example              JSON                       `json:"example,omitempty" protobuf:"bytes,36,rep,name=example"`
 }
 
 type JSON struct {
-	Raw []byte `json:"raw" protobuf:"bytes,1,opt,name=raw"`
+	Raw []byte `protobuf:"bytes,1,opt,name=raw"`
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: this is also `opt`, like in `Unknown`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go (hunk 16 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: `opt` for the other of course.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/deepcopy.go [hunk 18]
```diff
@@ -26,11 +26,13 @@ func (in *JSONSchemaProps) DeepCopy() *JSONSchemaProps {
 	*out = *in
 
 	if in.Default != nil {
-		out.Default = deepCopyJSON(in.Default)
+		defaultJSON := JSON(deepCopyJSON(*(in.Default)))
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/deepcopy.go]: you might want to set nil if in.Default is nil.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/deepcopy.go]: right :+1:

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go [hunk 19]
```diff
@@ -64,13 +64,15 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				}
 			}
 			if c.RandBool() {
-				obj.Default = `{"some": {"json": "test"}, "string": 42}` // some valid json
+				validJSON := apiextensions.JSON(`{"some": {"json": "test"}, "string": 42}`)
+				obj.Default = &validJSON
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go]: should be possible in one line as well

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go [hunk 20]
```diff
@@ -120,15 +120,16 @@ func (s *JSONSchemaPropsOrArray) UnmarshalJSON(data []byte) error {
...
+		return s.Raw, nil
 	}
 	return []byte("null"), nil
 
 }
 
 func (s *JSON) UnmarshalJSON(data []byte) error {
-	if err := json.Unmarshal(data, &(s.Raw)); err != nil {
-		return err
+	if len(data) > 0 && string(data[0]) != "null" {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: data[0] ? should be string(data), shoulnd't it?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: yes :+1:

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go [hunk 21]
```diff
@@ -120,15 +120,16 @@ func (s *JSONSchemaPropsOrArray) UnmarshalJSON(data []byte) error {
...
+		s.Raw = data
 	}
+	s = nil
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: this has no effect.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go [hunk 22]
```diff
@@ -139,3 +140,87 @@ type CustomResourceDefinitionList struct {
...
+	AdditionalProperties *JSONSchemaPropsOrBool
+	PatternProperties    map[string]JSONSchemaProps
+	Dependencies         JSONSchemaDependencies
+	AdditionalItems      *JSONSchemaPropsOrBool
+	Definitions          JSONSchemaDefinitions
+	ExternalDocs         *ExternalDocumentation
+	Example              *JSON
+}
+
+// JSON represents any valid JSON value.
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:190]: can you describe which types we support? `float64, bool, ..., map[string]interface{}, []interface{}`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go [hunk 23]
```diff
@@ -139,3 +140,88 @@ type CustomResourceDefinitionList struct {
...
+// These types are supported: bool, int64, float64, string, []interface{}, map[string]interface{} and nil.
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:191]: :+1:

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion_test.go [hunk 24]
```diff
@@ -63,11 +61,22 @@ func TestJSONConversion(t *testing.T) {
 		},
 	}
 
+	scheme := runtime.NewScheme()
+
+	// add internal and external types
+	if err := apiextensions.AddToScheme(scheme); err != nil {
+		t.Fatal(err)
+	}
+	if err := AddToScheme(scheme); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion_test.go:99]: 👍

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion_test.go [hunk 25]
```diff
@@ -32,31 +32,53 @@ func TestJSONConversion(t *testing.T) {
...
 		"null": {
-			input: &nullJSON,
-			expected: &JSON{
-				Raw: []byte(`null`),
+			input: &apiextensions.JSONSchemaProps{
+				Default: &nullJSON,
+			},
+			expected: &JSONSchemaProps{
+				Default: &JSON{
+					Raw: []byte(`null`),
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion_test.go]: here I would expect also `nil` in Default. We possibly have to set nil in that case in the custom conversion.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion_test.go]: btw, better call `nullJSON` `nilJSON`. Had to double-check that it's actually nil.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 26]
```diff
@@ -37,15 +37,20 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
...
 
 func Convert_apiextensions_JSON_To_v1beta1_JSON(in *apiextensions.JSON, out *JSON, s conversion.Scope) error {
-	if in != nil {
-		raw, err := json.Marshal(*in)
-		if err != nil {
-			return err
-		}
-		out.Raw = raw
-	} else {
+	if in == nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: can this happen? Isn't the `== nil` test done when this func is called?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 27]
```diff
@@ -37,15 +37,20 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
...
+		out = nil
+		return nil
+	}
+	if *in == nil {
 		out = nil
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: `*out = nil`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go [hunk 28]
```diff
@@ -36,16 +37,24 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
...
-		return nil
+func Convert_apiextensions_JSONSchemaProps_To_v1beta1_JSONSchemaProps(in *apiextensions.JSONSchemaProps, out *JSONSchemaProps, s conversion.Scope) error {
+	if err := autoConvert_apiextensions_JSONSchemaProps_To_v1beta1_JSONSchemaProps(in, out, s); err != nil {
+		return err
 	}
-	if *in == nil {
-		out = nil
-		return nil
+	if in.Default != nil {
+		if *(in.Default) == nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go]: nit: simpler with `&&`

**Pull request closed without merging**